site/_posts/2020-11-15-how-to-securely-implement-oauth-rails.md

```diff
@@ -0,0 +1,262 @@
+              client_secret: @client_secret,
+              redirect_uri: @redirect_uri }.to_query
+
+    response = RestClient.post("#{@identity_provider_url}oauth2/token?" << query, {})
+    token = JSON.parse(response.body)["access_token"]
+    session[:user_jwt] = {value: token, httponly: true}
+    redirect_to root_path
+  end
+
+  def destroy
+    reset_session
+    redirect_to root_path
+  end
+end
+```
+In summary, `oauth_callback` does a few things:
+1. It handles the redirect following FusionAuth authentication receiving an `authorization code`.
+2. It makes an additional call to exchange the authorization code for an access token.
+3. Saves the `access token` (JWT) on the session to indicate that the user has successfully authenticated.
+
+When a user clicks on the `Login` link, they will be presented with a login form from FusionAuth.
+Upon successful authentication, the FusionAuth redirects back to the Rails app along with an `authorization_code`. This is where we pick up in the `oauth_callback` method.
+
+The next step will be to exchange our authorization code for an access token.
+While we are authenticated, it is not sufficient access to access a user's resources. For that, we will make an additional [code exchange](https://fusionauth.io/docs/v1/tech/oauth/endpoints/#complete-the-authorization-code-grant-request) request. 
+
+We receive the access token encoded as a `JWT`. Finally, we save the JWT on the user session and redirect to the Welcome page. 
+
+The `logout` callback is much simpler. Similar to `oauth_callback` in the way that it receives the authorize redirect, our `destroy` method receives the logout redirect.
+Receiving the redirect tells us that the user has been logged out of FusionAuth and we are safe to clear the user's session in Rails.
+
+### Polishing up our Rails session
+Now that we have the functionality to authenticate users as well as log them out of FusionAuth, we will want to complete the same cycle on our Rails app.
+
+There are numerous ways to handle user sessions in Rails, but for this example, I decided to create helper methods accessible to all sub-classes of ApplicationController.
+
+`current_user` looks for a user jwt on the session. If it exists, it decodes the JWT and retrieves user's email address.
+If it does not exist, this means the user has logged out or their session has expired.
+
+```ruby
+# app/controllers/application_controller.rb
+
+class ApplicationController < ActionController::Base
+  helper_method :current_user
+  helper_method :logged_in?
+
+  def current_user
+    if session[:user_jwt]
+      token = session[:user_jwt]
+      decoded = TokenService.decode(token["value"])
+      decoded.first["email"]
+    end
+
+  end
+
+  def logged_in?
+    current_user.present?
+  end
+
+end
+```
+
+```ruby
+# app/services/token_service.rb
+
+class TokenService
+  def self.decode(token)
+    JWT.decode(
+        token,
+        nil,
+        false,
+        {algorithm: 'HS256'})
```

Comment:
where is the secret coming from? envt? config?

Also, it'd be good to check a few other claims:

* iss
* aud

And link out to https://fusionauth.io/learn/expert-advice/tokens/building-a-secure-jwt/